### PR TITLE
Suffix cache

### PR DESCRIPTION
**Pull request by apcj**
Experimental changes that could ultimately replace InFlightMap with BoundedArrayBackedRaftLog. Using the same RaftLog interface makes the API easy to reason about and reduces complexity, especially in the core raft machinery - since caching is a performance concern that must not affect behaviour, its details shouldn't leak into the core raft machinery.

### CODE REVIEW

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/core/consensus/log/segmented/CachedSuffixRaftLog.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        cacheLog.truncate( fromIndex );
+        fullLog.truncate( fromIndex );
+    }
+
+    @Override
+    public long prune( long safeIndex ) throws IOException
+    {
+        cacheLog.prune( safeIndex );
+        return fullLog.prune( safeIndex );
```
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/core/consensus/log/segmented/CachedSuffixRaftLog.java:59]: missing assertion here?

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/core/consensus/log/BoundedArrayBackedRaftLog.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+            throw new IllegalStateException( String.format( "Non-monotonic term %d for in entry %s in term %d",
+                    logEntry.term(), logEntry.toString(), term ) );
+        }
+
+        if ( appendIndex > prevIndex && offset( prevIndex + 1 ) == offset( appendIndex + 1 ) )
+        {
+            prune( prevIndex + 1 );
+        }
+        appendIndex++;
+        array[offset( appendIndex )] = logEntry;
```
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/core/consensus/log/BoundedArrayBackedRaftLog.java:80]: If pruning didn't work, this code will replace old entries with newer ones, is that correct?

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/core/consensus/log/BoundedArrayBackedRaftLog.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+/**
+ * In-memory raft log with a fixed maximum length, backed by an array.
+ *
+ * When entries are appended that would exceed this log's capacity, it discards old entries, with behaviour equivalent
+ * to {@link #prune(long)}.
+ * This behaviour is suitable for caching the suffix of a RaftLog, where all entries are duplicated in more durable
+ * underlying raft log. It is therefore fine to prune this log aggressively and for it to prune itself, because any
+ * discarded entries will still be present in the underlying log.
+ */
+public class BoundedArrayBackedRaftLog implements RaftLog
```
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/core/consensus/log/BoundedArrayBackedRaftLog.java:35]: This isn't used yet, correct?

**Comment by apcj** [comment]
Closing for now. More work required before ready for a merge.

**Pull request closed without merging**